imap/backupcyrusd.c

```diff
@@ -0,0 +1,757 @@
+#include "iostat.h"
+
+extern int optind;
+extern char *optarg;
+extern int opterr;
+
+static const char *bcd_clienthost = "[local]";
+static struct protstream *bcd_out = NULL;
+static struct protstream *bcd_in = NULL;
+static int bcd_logfd = -1;
+
+static struct proc_handle *proc_handle = NULL;
+
+/* signal to config.c */
+const int config_need_data = CONFIG_NEED_PARTITION_DATA;
+
+/* current namespace */
+static struct namespace bcd_namespace;
+
+/* Functions */
+static int do_fdata();
+static int do_fmeta();
+static int do_fmultistatus();
+static int do_meta();
+
+static void cmdloop(void);
+void shut_down(int code) __attribute__ ((noreturn));
+
+static void bcd_reset(void)
+{
+    uint64_t bytes_in = 0;
+    uint64_t bytes_out = 0;
+
+    proc_cleanup(&proc_handle);
+
+    if (bcd_in) {
+        prot_NONBLOCK(bcd_in);
+        prot_fill(bcd_in);
+        bytes_in = prot_bytes_in(bcd_in);
+        prot_free(bcd_in);
+    }
+
+    if (bcd_out) {
+        prot_flush(bcd_out);
+        bytes_out = prot_bytes_out(bcd_out);
+        prot_free(bcd_out);
+    }
+
+    auditlog_traffic(bytes_in, bytes_out);
+
+    bcd_in = bcd_out = NULL;
+
+    cyrus_reset_stdio();
+
+    bcd_clienthost = "[local]";
+    if (bcd_logfd != -1) {
+        close(bcd_logfd);
+        bcd_logfd = -1;
+    }
+
+    slowio_reset();
+}
+
+/*
+ * run once when process is forked;
+ * MUST NOT exit directly; must return with non-zero error code
+ */
+int service_init(int argc, char **argv,
+                 char **envp __attribute__((unused)))
+{
+    int r;
+
+    if (geteuid() == 0) fatal("must run as the Cyrus user", EX_USAGE);
+    proc_settitle_init(argc, argv, envp);
+
+    /* set signal handlers */
+    signals_set_shutdown(&shut_down);
+    signal(SIGPIPE, SIG_IGN);
+
+    /* Set namespace */
+    if ((r = mboxname_init_namespace(&bcd_namespace, 0))) {
+        syslog(LOG_ERR, "%s", error_message(r));
+        fatal(error_message(r), EX_CONFIG);
+    }
+
+    return 0;
+}
+
+/*
+ * run for each accepted connection
+ */
+int service_main(int argc __attribute__((unused)),
+                 char **argv __attribute__((unused)),
+                 char **envp __attribute__((unused)))
+{
+    const char *localip, *remoteip;
+
+    signals_poll();
+
+    bcd_in = prot_new(0, 0);
+    bcd_out = prot_new(1, 1);
+
+    /* Find out name of client host */
+    bcd_clienthost = get_clienthost(0, &localip, &remoteip);
+
+    /* Set inactivity timer */
+    prot_settimeout(bcd_in, 600);
+    prot_setflushonread(bcd_in, bcd_out);
+
+    cmdloop();
+
+    /* QUIT executed */
+
+    /* cleanup */
+    bcd_reset();
+
+    return 0;
+}
+
+/* Called by service API to shut down the service */
+void service_abort(int error)
+{
+    shut_down(error);
+}
+
+/*
+ * Cleanly shut down and exit
+ */
+void shut_down(int code)
+{
+    uint64_t bytes_in = 0;
+    uint64_t bytes_out = 0;
+
+    in_shutdown = 1;
+
+    libcyrus_run_delayed();
+
+    proc_cleanup(&proc_handle);
+
```

Comment:
Feels weird to duplicate all of this code from bcd_reset, should we refactor that somehow?

Reply:
Seems reasonable, though I'm not too concerned.  This is copy-paste from another daemon which did the same thing.